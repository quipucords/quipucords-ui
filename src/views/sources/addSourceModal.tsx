/**
 * Add Source Modal Component
 *
 * This component displays a modal for adding or editing a source of a specific type. It provides
 * a form to input source details including name, hosts, port, credential, and SSL settings.
 *
 * @module addSourceModal
 */
import React, { useCallback, useEffect, useState } from 'react';
import {
  ActionGroup,
  Button,
  Checkbox,
  Form,
  FormGroup,
  HelperText,
  HelperTextItem,
  Modal,
  ModalVariant,
  TextArea,
  TextInput
} from '@patternfly/react-core';
import { SimpleDropdown } from '../../components/simpleDropdown/simpleDropdown';
import { TypeaheadCheckboxes } from '../../components/typeAheadCheckboxes/typeaheadCheckboxes';
import { helpers } from '../../helpers';
import { useGetCredentialsApi } from '../../hooks/useCredentialApi';
import { type SourceType } from '../../types/types';

interface AddSourceModalProps {
  isOpen: boolean;
  source?: SourceType;
  sourceType?: string;
  onClose?: () => void;
  onSubmit?: (payload: any) => void;
}

interface SourceFormProps extends Omit<AddSourceModalProps, 'isOpen'> {
  useForm?: typeof useSourceForm;
}

interface SourceFormType {
  credentials?: number[];
  useParamiko?: boolean;
  sslVerify?: boolean;
  sslProtocol: string;
  name?: string;
  hosts?: string;
  port?: string;
}

const useSourceForm = ({
  sourceType,
  source,
  useGetCredentials = useGetCredentialsApi
}: { sourceType?: string; source?: Partial<SourceType>; useGetCredentials?: typeof useGetCredentialsApi } = {}) => {
  const initialFormState: SourceFormType = {
    credentials: [],
    useParamiko: false,
    sslVerify: true,
    sslProtocol: 'SSLv23',
    name: '',
    hosts: undefined,
    port: ''
  };

  const { getCredentials } = useGetCredentials();
  const [credOptions, setCredOptions] = useState<{ value: string; label: string }[] | []>([]);
  const [formData, setFormData] = useState<SourceFormType>(initialFormState);

  const typeValue = source?.source_type || sourceType?.split(' ')?.shift()?.toLowerCase();
  const isNetwork = typeValue === 'network';
  const isOpenshift = typeValue === 'openshift';

  // Edit props, reset state on unmount
  useEffect(() => {
    if (source) {
      setFormData({
        credentials: source?.credentials?.map(c => c.id) || [],
        useParamiko: source?.options?.use_paramiko || false,
        sslVerify: source?.options?.ssl_cert_verify ?? true,
        sslProtocol: (source?.options?.disable_ssl && 'Disable SSL') || source?.options?.ssl_protocol || 'SSLv23',
        name: source?.name || '',
        hosts: source?.hosts?.join(',') || '',
        port: source?.port?.toString() || ''
      });
    }

    return () => {
      setFormData(initialFormState);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    getCredentials({
      params: {
        cred_type: typeValue
      }
    })
      .then(response => {
        const updatedOptions = response?.data?.results?.map(({ name, id }) => ({ label: name, value: `${id}` }));
        setCredOptions(updatedOptions || []);
      })
      .catch(err => {
        if (!helpers.TEST_MODE) {
          console.error(err);
        }
      });
  }, [typeValue, getCredentials]);

  const handleInputChange = useCallback(
    (field: string, value: unknown) => {
      setFormData({ ...formData, [field]: value });
    },
    [formData]
  );

  const filterFormData = useCallback(
    (data = formData) => {
      const { credentials, useParamiko, sslVerify, sslProtocol, name, hosts, port } = data;
      return {
        name: name,
        credentials: credentials?.map(c => Number(c)),
        hosts: helpers.normalizeHosts(hosts),
        port: port || (isOpenshift && '6443') || (isNetwork && '22') || '443',
        options: !isNetwork
          ? {
              ssl_cert_verify: sslProtocol !== 'Disable SSL' && sslVerify,
              disable_ssl: sslProtocol === 'Disable SSL',
              ...(sslProtocol !== 'Disable SSL' && { ssl_protocol: sslProtocol })
            }
          : {
              use_paramiko: useParamiko
            },
        ...(!source && { source_type: typeValue }),
        ...(source && { id: source.id })
      };
    },
    [isNetwork, isOpenshift, formData, source, typeValue]
  );

  return {
    credOptions,
    formData,
    isNetwork,
    isOpenshift,
    handleInputChange,
    filterFormData,
    typeValue
  };
};

const SourceForm: React.FC<SourceFormProps> = ({
  source,
  sourceType,
  onClose = () => {},
  onSubmit = () => {},
  useForm = useSourceForm
}) => {
  const { formData, isNetwork, isOpenshift, credOptions, handleInputChange, filterFormData } = useForm({
    sourceType,
    source
  });
  const onAdd = () => onSubmit(filterFormData());

  return (
    <Form>
      <FormGroup label="Name" isRequired fieldId="name">
        <TextInput
          value={formData?.name}
          placeholder="Enter a name for the source"
          isRequired
          type="text"
          id="source-name"
          name="name"
          onChange={event => handleInputChange('name', (event.target as HTMLInputElement).value)}
          ouiaId="name"
        />
      </FormGroup>
      <FormGroup label="Credentials" fieldId="credentials" isRequired>
        <TypeaheadCheckboxes
          onChange={(selections: string[]) => {
            const validIds = selections.map(Number).filter(id => !isNaN(id));
            handleInputChange('credentials', validIds);
          }}
          options={credOptions}
          selectedOptions={formData?.credentials?.map(String) || []}
          menuToggleOuiaId="add_credentials_select"
          maxSelections={isNetwork ? Infinity : 1} // Limit selection to 1 for non-network sources
        />
        {!isNetwork && (
          <HelperText>
            <HelperTextItem variant="warning">Only one credential can be selected for this source type.</HelperTextItem>
          </HelperText>
        )}
      </FormGroup>
      {isNetwork ? (
        <React.Fragment>
          <FormGroup label="Search addresses" isRequired fieldId="hosts">
            <TextArea
              placeholder="Enter values separated by commas"
              value={formData?.hosts}
              onChange={event => handleInputChange('hosts', event.target.value)}
              validated={helpers.validateHosts(formData?.hosts, Infinity)}
              isRequired
              id="source-hosts"
              name="hosts"
              data-ouia-component-id="hosts_multiple"
            />
            <HelperText>
              Type IP addresses, IP ranges, and DNS host names. Use CIDR or Ansible notation for ranges.
            </HelperText>
          </FormGroup>
          <FormGroup label="Port" fieldId="port">
            <TextInput
              value={formData?.port}
              placeholder="Optional"
              type="text"
              id="source-port"
              name="port"
              onChange={event => handleInputChange('port', (event.target as HTMLInputElement).value)}
              ouiaId="port"
            />
            <HelperText>Default port is 22</HelperText>
          </FormGroup>
        </React.Fragment>
      ) : (
        <React.Fragment>
          <FormGroup label="IP address or hostname" isRequired fieldId="hosts">
            <TextInput
              value={formData?.hosts}
              onChange={event => handleInputChange('hosts', (event.target as HTMLInputElement).value)}
              isRequired
              validated={helpers.validateHosts(formData?.hosts, 1)}
              id="source-hosts"
              name="hosts"
              ouiaId="hosts_single"
            />
            <HelperText>Enter an IP address or hostname</HelperText>
          </FormGroup>
          <FormGroup label="Port" fieldId="port">
            <TextInput
              value={formData?.port}
              placeholder="Optional"
              type="text"
              id="source-port"
              name="port"
              onChange={event => handleInputChange('port', (event.target as HTMLInputElement).value)}
              ouiaId="port"
            />
            <HelperText>Default port is {isOpenshift ? '6443' : '443'}</HelperText>
          </FormGroup>
        </React.Fragment>
      )}
      {isNetwork ? (
        <FormGroup label="" fieldId="paramiko">
          <Checkbox
            key="paramiko"
            label="Connect using Paramiko instead of Open SSH"
            id="paramiko"
            isChecked={formData?.useParamiko}
            onChange={(_ev, checked) => handleInputChange('useParamiko', checked)}
            ouiaId="options_paramiko"
          />
        </FormGroup>
      ) : (
        <React.Fragment>
          <FormGroup label="Connection" fieldId="connection">
            <SimpleDropdown
              isFullWidth
              label={formData?.sslProtocol || 'Select protocol'}
              menuToggleOuiaId="options_ssl_protocol"
              variant={'default'}
              onSelect={item => handleInputChange('sslProtocol', item)}
              dropdownItems={[
                { item: 'SSLv23', ouiaId: 'sslv23' },
                { item: 'TLSv1', ouiaId: 'tlsv1' },
                { item: 'TLSv1.1', ouiaId: 'tlsv11' },
                { item: 'TLSv1.2', ouiaId: 'tlsv12' },
                { item: 'Disable SSL', ouiaId: 'disable_ssl' }
              ]}
            />
          </FormGroup>
          <FormGroup label="" fieldId="ssl_verify">
            <Checkbox
              key="ssl_verify"
              label="Verify SSL certificate"
              id="ssl_verify"
              isDisabled={formData?.sslProtocol === 'Disable SSL'}
              isChecked={formData?.sslProtocol !== 'Disable SSL' && formData?.sslVerify}
              onChange={(_ev, checked) => handleInputChange('sslVerify', checked)}
              ouiaId="options_ssl_cert"
            />
          </FormGroup>
        </React.Fragment>
      )}
      <ActionGroup>
        <Button variant="primary" onClick={onAdd}>
          Save
        </Button>
        <Button variant="link" onClick={() => onClose()}>
          Cancel
        </Button>
      </ActionGroup>
    </Form>
  );
};

const AddSourceModal: React.FC<AddSourceModalProps> = ({
  isOpen,
  source,
  sourceType,
  onClose = () => {},
  onSubmit = () => {}
}) => (
  <Modal
    variant={ModalVariant.small}
    title={(source && `Edit Source: ${source.name || ''}`) || `Add Source: ${sourceType || ''}`}
    isOpen={isOpen}
    onClose={() => onClose()}
  >
    <SourceForm source={source} sourceType={sourceType} onClose={onClose} onSubmit={onSubmit} />
  </Modal>
);

export { AddSourceModal as default, AddSourceModal, SourceForm, useSourceForm, type AddSourceModalProps };
