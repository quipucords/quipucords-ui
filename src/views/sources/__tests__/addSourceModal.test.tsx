import React, { act } from 'react';
import { render, renderHook, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import { shallowComponent } from '../../../../config/jest.setupTests';
import { AddSourceModal, SourceForm, useSourceForm } from '../addSourceModal';

describe('AddSourceModal', () => {
  let mockOnClose;
  let mockOnSubmit;

  beforeEach(async () => {
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({}));

    mockOnClose = jest.fn();
    mockOnSubmit = jest.fn();
    await act(() => {
      render(
        <AddSourceModal
          isOpen={true}
          sourceType="network"
          source={undefined}
          onClose={mockOnClose}
          onSubmit={mockOnSubmit}
        />
      );
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const component = await shallowComponent(<AddSourceModal isOpen={true} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should have the correct title', () => {
    const title = screen.getByText(/Add\sSource:\snetwork/i);
    expect(title).toMatchSnapshot('title');
  });

  it('should call onSubmit with the correct filtered data when "Save" is clicked', async () => {
    const user = userEvent.setup();
    await user.type(screen.getByPlaceholderText('Enter a name for the source'), 'Test Source');
    await user.click(screen.getByText('Save'));

    expect(mockOnSubmit.mock.calls).toMatchSnapshot('onSubmit, filtered data');
  });

  it('should call onClose', async () => {
    const user = userEvent.setup();
    await user.click(screen.getByText('Cancel'));

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });
});

describe('useSourceForm', () => {
  beforeEach(() => {
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({}));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should initialize formData correctly', async () => {
    const result = renderHook(() => useSourceForm({ sourceType: 'network' })).result;
    await act(() => expect(result.current.formData).toMatchSnapshot('formData'));
  });

  it('should allow editing a source', async () => {
    const result = renderHook(() =>
      useSourceForm({
        source: {
          id: 123,
          name: 'lorem',
          port: 456,
          source_type: 'openshift'
        }
      })
    ).result;

    await act(() => expect(result.current.formData).toMatchSnapshot('formData, edit'));
  });

  it('should update and filter formData when handleInputChange is called', async () => {
    const result = renderHook(() => useSourceForm({ sourceType: 'network' })).result;
    const mockValue = 'Lorem ipsum';

    await act(() => result.current.handleInputChange('name', mockValue));
    expect(result.current.formData.name).toBe(mockValue);

    const filteredData = result.current.filterFormData();
    expect(filteredData.name).toBe(mockValue);
  });
});

describe('SourceForm', () => {
  it('should render a basic component', async () => {
    const component = await shallowComponent(<SourceForm />);
    expect(component).toMatchSnapshot('basic');
  });
});
