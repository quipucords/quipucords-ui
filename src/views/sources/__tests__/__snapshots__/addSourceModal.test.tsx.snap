// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSourceModal should call onSubmit with the correct filtered data when "Save" is clicked: onSubmit, filtered data 1`] = `
[
  [
    {
      "credentials": [],
      "hosts": [],
      "name": "Test Source",
      "options": {
        "use_paramiko": false,
      },
      "port": "22",
      "source_type": "network",
    },
  ],
]
`;

exports[`AddSourceModal should have the correct title: title 1`] = `
<span
  class="pf-v5-c-modal-box__title-text"
>
  Add Source: network
</span>
`;

exports[`AddSourceModal should render a basic component: basic 1`] = `
<Modal
  actions={[]}
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  position="default"
  showClose={true}
  title="Add Source: "
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <SourceForm
    onClose={[Function]}
    onSubmit={[Function]}
  />
</Modal>
`;

exports[`SourceForm should render a basic component: basic 1`] = `
<Form>
  <FormGroup
    fieldId="name"
    isRequired={true}
    label="Name"
  >
    <TextInput
      id="source-name"
      isRequired={true}
      name="name"
      onChange={[Function]}
      ouiaId="name"
      placeholder="Enter a name for the source"
      type="text"
      value=""
    />
  </FormGroup>
  <FormGroup
    fieldId="credentials"
    isRequired={true}
    label="Credentials"
  >
    <TypeaheadCheckboxes
      maxSelections={1}
      menuToggleOuiaId="add_credentials_select"
      onChange={[Function]}
      options={[]}
      selectedOptions={[]}
    />
    <HelperText>
      <HelperTextItem
        variant="warning"
      >
        Only one credential can be selected for this source type.
      </HelperTextItem>
    </HelperText>
  </FormGroup>
  <React.Fragment>
    <FormGroup
      fieldId="hosts"
      isRequired={true}
      label="IP address or hostname"
    >
      <TextInput
        id="source-hosts"
        isRequired={true}
        name="hosts"
        onChange={[Function]}
        ouiaId="hosts_single"
        validated="error"
        value=""
      />
      <HelperText>
        Enter an IP address or hostname
      </HelperText>
    </FormGroup>
    <FormGroup
      fieldId="port"
      label="Port"
    >
      <TextInput
        id="source-port"
        name="port"
        onChange={[Function]}
        ouiaId="port"
        placeholder="Optional"
        type="text"
        value=""
      />
      <HelperText>
        Default port is 
        443
      </HelperText>
    </FormGroup>
  </React.Fragment>
  <React.Fragment>
    <FormGroup
      fieldId="connection"
      label="Connection"
    >
      <SimpleDropdown
        dropdownItems={
          [
            {
              "item": "SSLv23",
              "ouiaId": "sslv23",
            },
            {
              "item": "TLSv1",
              "ouiaId": "tlsv1",
            },
            {
              "item": "TLSv1.1",
              "ouiaId": "tlsv11",
            },
            {
              "item": "TLSv1.2",
              "ouiaId": "tlsv12",
            },
            {
              "item": "Disable SSL",
              "ouiaId": "disable_ssl",
            },
          ]
        }
        isFullWidth={true}
        label="SSLv23"
        menuToggleOuiaId="options_ssl_protocol"
        onSelect={[Function]}
        variant="default"
      />
    </FormGroup>
    <FormGroup
      fieldId="ssl_verify"
      label=""
    >
      <Checkbox
        className=""
        id="ssl_verify"
        isChecked={true}
        isDisabled={false}
        isLabelWrapped={false}
        isRequired={false}
        isValid={true}
        label="Verify SSL certificate"
        onChange={[Function]}
        ouiaId="options_ssl_cert"
        ouiaSafe={true}
      />
    </FormGroup>
  </React.Fragment>
  <ActionGroup>
    <Button
      onClick={[Function]}
      variant="primary"
    >
      Save
    </Button>
    <Button
      onClick={[Function]}
      variant="link"
    >
      Cancel
    </Button>
  </ActionGroup>
</Form>
`;

exports[`SourceForm should render specifics to different source types: form, ansible 1`] = `
<div
  class="pf-v5-c-helper-text"
>
  Default port is 
  443
</div>
`;

exports[`SourceForm should render specifics to different source types: form, network 1`] = `
<div
  class="pf-v5-c-helper-text"
>
  Default port is 22
</div>
`;

exports[`SourceForm should render specifics to different source types: form, openshift 1`] = `
<div
  class="pf-v5-c-helper-text"
>
  Default port is 
  6443
</div>
`;

exports[`SourceForm should render specifics to different source types: form, rhacs 1`] = `
<div
  class="pf-v5-c-helper-text"
>
  Default port is 
  443
</div>
`;

exports[`SourceForm should render specifics to different source types: form, satellite 1`] = `
<div
  class="pf-v5-c-helper-text"
>
  Default port is 
  443
</div>
`;

exports[`SourceForm should render specifics to different source types: form, vcenter 1`] = `
<div
  class="pf-v5-c-helper-text"
>
  Default port is 
  443
</div>
`;

exports[`useSourceForm should allow editing a source: formData, edit 1`] = `
{
  "credentials": [],
  "hosts": "",
  "name": "lorem",
  "port": "456",
  "sslProtocol": "SSLv23",
  "sslVerify": true,
  "useParamiko": false,
}
`;

exports[`useSourceForm should initialize formData correctly: formData 1`] = `
{
  "credentials": [],
  "hosts": "",
  "name": "",
  "port": "",
  "sslProtocol": "SSLv23",
  "sslVerify": true,
  "useParamiko": false,
}
`;
